implicit/als.py

```diff
@@ -105,12 +105,36 @@ def recommend(self, userid, user_items, N=10, filter_items=None, recalculate_use
     def _user_factor(self, userid, user_items, recalculate_user=False):
         if not recalculate_user:
             return self.user_factors[userid]
-        Y = self.item_factors
-        if self._YtY is None:
-            self._YtY = Y.T.dot(Y)
-        return user_factor(Y, self._YtY, user_items, userid,
+        return user_factor(self.item_factors, self.YtY, user_items, userid,
                            self.regularization, self.factors)
 
+    def explain(self, userid, user_items, itemid, user_weights=None, N=10):
+        """ Returns the predicted rating for an user x item pair,
+            the explanation (the contribution from the top N items the user liked),
+            and a user latent factor weight that can be cached if you want to
+            get more than one explanation for the same user.
+        """
+        if user_weights is None:
+            A, _ = user_linear_equation(self.item_factors, self.YtY,
+                                        user_items, userid,
+                                        self.regularization, self.factors)
+            user_weights = np.linalg.inv(A)
```

Comment:
I would tweak this a little, and just set ```user_weights, _ = user_linear_equation ..``` and then ```weighted_item = np.linalg.solve(A, seed_item)```. Rationale behind this is here https://www.johndcook.com/blog/2010/01/19/dont-invert-that-matrix/